aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,602 @@
+                    oT,
+                    oH,
+                    oW});
+  } else {
+    output.resize_(
+        {self.size(0), oT, oH, oW});
+  }
+  output.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES(
+      self.type(), "max_unpooling3d_forward_out_cpu_frame", ([&] {
+        max_unpooling3d_forward_out_cpu_frame<scalar_t>(
+            output,
+            self.contiguous(),
+            indices.contiguous(),
+            oT,
+            oW,
+            oH,
+            stride[0],
+            stride[1],
+            stride[2],
+            padding[0],
+            padding[1],
+            padding[2]);
+      }));
+  return output;
+}
+
+Tensor max_unpooling3d_forward_cpu(
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size,
+    IntList stride,
+    IntList padding) {
+  auto output = at::empty({0}, self.options());
+  max_unpooling3d_forward_out_cpu(
+      output, self, indices, output_size, stride, padding);
+  return output;
+}
+
+template <typename scalar_t>
+static void max_unpooling2d_backward_out_cpu_frame(
+    scalar_t* gradInput_p,
+    scalar_t* gradOutput_p,
+    int64_t* ind_p,
+    int64_t nslices,
+    int64_t iwidth,
+    int64_t iheight,
+    int64_t owidth,
+    int64_t oheight) {
+  bool has_error = false;
+  int64_t error_index = 0;
+  int k;
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++) {
+    scalar_t* gradInput_p_k = gradInput_p + k * iwidth * iheight;
+    scalar_t* gradOutput_p_k = gradOutput_p + k * owidth * oheight;
+    int64_t* ind_p_k = ind_p + k * iwidth * iheight;
+
+    int64_t i, j;
+    int64_t maxp;
+
+    for (i = 0; i < iheight; i++) {
+      for (j = 0; j < iwidth; j++) {
+        maxp = ind_p_k[i * iwidth + j]; /* retrieve position of max */
+        if (maxp < 0 || maxp >= owidth * oheight) {
+          has_error = true;
+          error_index = maxp;
+        }
+        gradInput_p_k[i * iwidth + j] =
+            gradOutput_p_k[maxp]; /* update gradient */
+      }
+    }
+  }
+  if (has_error) {
+    AT_ERROR(
+        "invalid max index ",
+        error_index,
+        ", owidth= ",
+        owidth,
+        ", oheight= ",
+        oheight);
+  }
+}
+
+Tensor& max_unpooling2d_backward_out_cpu(
+    Tensor& grad_input,
+    const Tensor& grad_output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  int64_t owidth = output_size[0];
+  int64_t oheight = output_size[1];
+  int dimw = 2;
+  int dimh = 1;
+  int nbatch = 1;
+  int nslices;
+  int iheight;
+  int iwidth;
+  AT_CHECK(
+      indices.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type Long");
+  AT_CHECK(
+      self.sizes() == indices.sizes(), "Input shape must match indices shape");
+
+  AT_CHECK(output_size.size() == 2, "Output size must be 2");
+
+  /* get contiguous gradOutput and indices */
+  auto gradOutput = grad_output.contiguous();
+  auto indicesContiguous = indices.contiguous();
```

Comment:
I prefer it if you call the non-contiguous inputs `grad_output_` and `indices_` and the contiguous versions `grad_output` and `indices`, so you are less likely to mess up and use the wrong one in the code below...